In [1]:
import pandas as pd
import numpy as np


In [2]:
df_shared_articles = pd.read_csv('shared_articles.csv')

In [3]:
print(df_shared_articles.head())

    timestamp        eventType            contentId       authorPersonId  \
0  1459192779  CONTENT REMOVED -6451309518266745024  4340306774493623681   
1  1459193988   CONTENT SHARED -4110354420726924665  4340306774493623681   
2  1459194146   CONTENT SHARED -7292285110016212249  4340306774493623681   
3  1459194474   CONTENT SHARED -6151852268067518688  3891637997717104548   
4  1459194497   CONTENT SHARED  2448026894306402386  4340306774493623681   

       authorSessionId authorUserAgent authorRegion authorCountry contentType  \
0  8940341205206233829             NaN          NaN           NaN        HTML   
1  8940341205206233829             NaN          NaN           NaN        HTML   
2  8940341205206233829             NaN          NaN           NaN        HTML   
3 -1457532940883382585             NaN          NaN           NaN        HTML   
4  8940341205206233829             NaN          NaN           NaN        HTML   

                                                 url  \


In [4]:
print(df_shared_articles.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3122 entries, 0 to 3121
Data columns (total 13 columns):
timestamp          3122 non-null int64
eventType          3122 non-null object
contentId          3122 non-null int64
authorPersonId     3122 non-null int64
authorSessionId    3122 non-null int64
authorUserAgent    680 non-null object
authorRegion       680 non-null object
authorCountry      680 non-null object
contentType        3122 non-null object
url                3122 non-null object
title              3122 non-null object
text               3122 non-null object
lang               3122 non-null object
dtypes: int64(4), object(9)
memory usage: 317.2+ KB
None


In [5]:
print(df_shared_articles.describe())

          timestamp     contentId  authorPersonId  authorSessionId
count  3.122000e+03  3.122000e+03    3.122000e+03     3.122000e+03
mean   1.468804e+09  3.284110e+16    3.984819e+17     1.976675e+17
std    7.571512e+06  5.383734e+18    4.387115e+18     5.403753e+18
min    1.459193e+09 -9.222795e+18   -9.120686e+18    -9.212055e+18
25%    1.462369e+09 -4.660869e+18   -1.578288e+18    -4.651280e+18
50%    1.467136e+09  4.436584e+16   -7.092877e+17     4.154103e+17
75%    1.473873e+09  4.745617e+18    3.609194e+18     4.841644e+18
max    1.488308e+09  9.222265e+18    9.210531e+18     9.221043e+18


In [6]:
print(len(df_shared_articles))

3122


In [7]:
content_shared = df_shared_articles[df_shared_articles['eventType'] == 'CONTENT SHARED']
content_removed = df_shared_articles[df_shared_articles['eventType'] == 'CONTENT REMOVED']

In [8]:
print((len(content_shared)/len(df_shared_articles))*100,(len(content_removed)/len(df_shared_articles))*100)

97.59769378603458 2.4023062139654066


# 97.59% data is shared and only 2.40% data was removed

In [9]:
types_lang = df_shared_articles['lang'].unique()
print(types_lang)

['en' 'pt' 'es' 'la' 'ja']


In [10]:
dummy = [0,0,0,0,0]
dummy[0] = len(content_shared[content_shared['lang'] == 'en'])
dummy[1] = len(content_shared[content_shared['lang'] == 'pt'])
dummy[2] = len(content_shared[content_shared['lang'] == 'es']) 
dummy[3] = len(content_shared[content_shared['lang'] == 'la'])
dummy[4] = len(content_shared[content_shared['lang'] == 'ja'])
len_content_shared = len(content_shared)       
    

In [11]:
print(dummy[0]/len_content_shared *100)
print(dummy[1]/len_content_shared *100)
print(dummy[2]/len_content_shared *100)
print(dummy[3]/len_content_shared *100)
print(dummy[4]/len_content_shared *100)


72.56317689530685
27.207088939940927
0.06563833278634722
0.09845749917952085
0.06563833278634722


# about 99.7% shared articles belongs to lang as "en","pt"

In [12]:
df_user_interaction = pd.read_csv('users_interactions.csv')

In [13]:
print(df_user_interaction.head())

    timestamp eventType            contentId             personId  \
0  1465413032      VIEW -3499919498720038879 -8845298781299428018   
1  1465412560      VIEW  8890720798209849691 -1032019229384696495   
2  1465416190      VIEW   310515487419366995 -1130272294246983140   
3  1465413895    FOLLOW   310515487419366995   344280948527967603   
4  1465412290      VIEW -7820640624231356730  -445337111692715325   

             sessionId                                          userAgent  \
0  1264196770339959068                                                NaN   
1  3621737643587579081  Mozilla/5.0 (Macintosh; Intel Mac OS X 10_11_2...   
2  2631864456530402479                                                NaN   
3 -3167637573980064150                                                NaN   
4  5611481178424124714                                                NaN   

  userRegion userCountry  
0        NaN         NaN  
1         NY          US  
2        NaN         NaN  
3        NaN  

In [14]:
print(df_user_interaction.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 72312 entries, 0 to 72311
Data columns (total 8 columns):
timestamp      72312 non-null int64
eventType      72312 non-null object
contentId      72312 non-null int64
personId       72312 non-null int64
sessionId      72312 non-null int64
userAgent      56918 non-null object
userRegion     56907 non-null object
userCountry    56918 non-null object
dtypes: int64(4), object(4)
memory usage: 4.4+ MB
None


In [15]:
df_user_interaction['eventType'].unique()

array(['VIEW', 'FOLLOW', 'BOOKMARK', 'LIKE', 'COMMENT CREATED'],
      dtype=object)

In [16]:
dummy = [0,0,0,0,0]
dummy[0] = len(df_user_interaction[df_user_interaction['eventType'] == 'VIEW'])
dummy[1] = len(df_user_interaction[df_user_interaction['eventType'] == 'FOLLOW'])
dummy[2] = len(df_user_interaction[df_user_interaction['eventType'] == 'BOOKMARK']) 
dummy[3] = len(df_user_interaction[df_user_interaction['eventType'] == 'LIKE'])
dummy[4] = len(df_user_interaction[df_user_interaction['eventType'] == 'COMMENT CREATED'])
len_df_user_interactions = len(df_user_interaction)       
    

In [17]:
print(dummy[0]/len_df_user_interactions *100)
print(dummy[1]/len_df_user_interactions *100)
print(dummy[2]/len_df_user_interactions *100)
print(dummy[3]/len_df_user_interactions *100)
print(dummy[4]/len_df_user_interactions *100)


84.47560570859608
1.945735147693329
3.4060736807168936
7.944739462329904
2.2278460006637903


# about 84.4% articles were viewed followed by 15.6% other activities.

In [18]:
event_type_strength = {
   'VIEW': 1.0,
   'LIKE': 2.0, 
   'BOOKMARK': 2.5, 
   'FOLLOW': 3.0,
   'COMMENT CREATED': 4.0,  
}

In [19]:
df_user_interaction['eventStrength'] = df_user_interaction['eventType'].apply(lambda x: event_type_strength[x])

In [20]:
print(len(df_user_interaction['personId'].unique()))
print(len(df_user_interaction['contentId'].unique()))
print(len(df_user_interaction['contentId']))

1895
2987
72312


In [21]:
duplicate_df_user_interaction = df_user_interaction[['contentId','eventStrength']]
for_graph = duplicate_df_user_interaction.groupby(['contentId'])['eventStrength'].sum().reset_index()
print(for_graph['eventStrength'].mean())


29.92919317040509


In [22]:
import seaborn as sns
x1 = for_graph['contentId']
y1 = for_graph['eventStrength']
sns.scatterplot(x=x1,y=y1)

In [23]:
df_users_interactions_count = df_user_interaction.groupby(['personId', 'contentId']).size().groupby('personId').size()

In [24]:
df_users_with_enough_interactions = df_users_interactions_count[df_users_interactions_count >= 5].reset_index()[['personId']]

In [25]:
df_interactions_from_selected_users = df_user_interaction.merge(df_users_with_enough_interactions,how = 'right',left_on = 'personId',right_on = 'personId')

In [26]:
df_interactions_from_selected_users = df_interactions_from_selected_users[['contentId','personId','eventStrength']]

In [27]:
print(df_interactions_from_selected_users)

                 contentId             personId  eventStrength
0     -3499919498720038879 -8845298781299428018            1.0
1     -3499919498720038879 -8845298781299428018            1.0
2      6437568358552101410 -8845298781299428018            1.0
3     -4760639635023250284 -8845298781299428018            1.0
4     -6142462826726347616 -8845298781299428018            1.0
...                    ...                  ...            ...
69863 -1876558460087500339 -9063420486253202900            2.5
69864  -912681925384819247 -9063420486253202900            2.5
69865 -1419188045265393093 -9063420486253202900            1.0
69866 -1419188045265393093 -9063420486253202900            1.0
69867 -1419188045265393093 -9063420486253202900            1.0

[69868 rows x 3 columns]


In [28]:
df_final_interactions = df_interactions_from_selected_users.groupby(['personId', 'contentId'])['eventStrength'].sum().reset_index()

In [29]:
print(df_final_interactions)

                  personId            contentId  eventStrength
0     -9223121837663643404 -8949113594875411859            1.0
1     -9223121837663643404 -8377626164558006982            1.0
2     -9223121837663643404 -8208801367848627943            1.0
3     -9223121837663643404 -8187220755213888616            1.0
4     -9223121837663643404 -7423191370472335463            8.0
...                    ...                  ...            ...
39101  9210530975708218054  8477804012624580461            8.5
39102  9210530975708218054  8526042588044002101            1.0
39103  9210530975708218054  8856169137131817223            1.0
39104  9210530975708218054  8869347744613364434            1.0
39105  9210530975708218054  9209886322932807692            1.0

[39106 rows x 3 columns]


In [30]:
print(len(df_final_interactions['personId'].unique()))
print(len(df_final_interactions['contentId'].unique()))



1140
2984


# Train Test split

In [31]:
from sklearn.model_selection import train_test_split
df_train_interactions, df_test_interactions = train_test_split(df_final_interactions,
                                   stratify=df_final_interactions['personId'], 
                                   test_size=0.20,
                                   random_state=42)

print('# interactions on Train set: %d' % len(df_train_interactions))
print('# interactions on Test set: %d' % len(df_test_interactions))

# interactions on Train set: 31284
# interactions on Test set: 7822


# popularity model

In [32]:
#Computes the most popular items
df_popularity_model = df_final_interactions.groupby('contentId')['eventStrength'].sum().sort_values(ascending=False).reset_index()
df_popularity_model.head(10)

contentId  eventStrength
0 -4029704725707465084          457.5
1 -2358756719610361882          361.0
2 -6783772548752091658          357.0
3  8657408509986329668          338.0
4  -133139342397538859          336.5
5 -8208801367848627943          327.5
6 -6843047699859121724          315.0
7  2581138407738454418          310.0
8  2857117417189640073          308.5
9 -1633984990770981161          298.0

# content based model

# note that tf idf is used on title+text of given data

In [33]:
dup_df_shared_articles = df_shared_articles
dup_df_shared_articles['title_and_text'] = df_shared_articles['title'] + " " + df_shared_articles['text']


In [34]:
from nltk.corpus import stopwords
from sklearn.feature_extraction.text import TfidfVectorizer
stopwords_list = stopwords.words('english') + stopwords.words('portuguese')
vectorizer = TfidfVectorizer(analyzer='word',
                     token_pattern=r'\w{1,}',
                     ngram_range=(1, 2),
                     min_df=0.003,
                     max_df=0.5,
                     max_features=5000,
                     stop_words=stopwords_list)
tfidf_matrix = vectorizer.fit_transform((dup_df_shared_articles['title_and_text']).str.lower())


In [35]:
print(len(dup_df_shared_articles['title_and_text']))
print(tfidf_matrix.shape)
print(type(tfidf_matrix))
#print(tfidf_matrix)

3122
(3122, 5000)
<class 'scipy.sparse.csr.csr_matrix'>


In [36]:
print(tfidf_matrix[0])

  (0, 16)	0.023412764268819004
  (0, 948)	0.02354951449797743
  (0, 368)	0.02104876375343444
  (0, 4204)	0.022492824577581432
  (0, 4186)	0.029546127991827605
  (0, 4972)	0.023762037789310526
  (0, 2528)	0.04682552853763801
  (0, 2721)	0.023345787838616357
  (0, 2394)	0.02272182827159206
  (0, 3721)	0.0204968448625716
  (0, 2529)	0.01745237523818198
  (0, 334)	0.014775675894734566
  (0, 153)	0.018095378123662804
  (0, 1307)	0.01710337696674317
  (0, 1249)	0.02383496788887259
  (0, 4973)	0.011886073287213132
  (0, 14)	0.012118475456803652
  (0, 2610)	0.014361948784163242
  (0, 1885)	0.022663570885372627
  (0, 29)	0.010543414090243975
  (0, 736)	0.019280463538296252
  (0, 3531)	0.015201885597581028
  (0, 4513)	0.013663230680366978
  (0, 4396)	0.01371938366583984
  (0, 711)	0.024378177352018415
  :	:
  (0, 4014)	0.016262339249444705
  (0, 3964)	0.04006977097971936
  (0, 2748)	0.059069407987017646
  (0, 3535)	0.04415307897058428
  (0, 2569)	0.02104876375343444
  (0, 4465)	0.016588464548583

In [37]:
tfidf_feature_names = vectorizer.get_feature_names()

In [38]:
print(len(tfidf_feature_names))

5000


In [39]:
from sklearn.metrics.pairwise import cosine_similarity
from sklearn.metrics.pairwise import sigmoid_kernel

In [40]:
sigmoid_kernel_output = sigmoid_kernel(tfidf_matrix,tfidf_matrix)

In [41]:
print(type(sigmoid_kernel_output))
print(sigmoid_kernel_output.shape)


<class 'numpy.ndarray'>
(3122, 3122)


In [42]:
print(sigmoid_kernel_output[0])

[0.76167814 0.76167814 0.76160088 ... 0.76159991 0.76160459 0.76159557]


In [43]:
dic_dup_df_shared_articles = dup_df_shared_articles['title_and_text'].to_dict()

In [44]:
key_list = list(dic_dup_df_shared_articles.keys()) 
val_list = list(dic_dup_df_shared_articles.values()) 

In [45]:
def gives_results(input_,sigmoid_kernel_output):
    index = key_list[val_list.index(input_)]
    content_sigmoid_output = list(enumerate(sigmoid_kernel_output[index]))
    content_sigmoid_output1 = sorted(content_sigmoid_output, key=lambda x: x[1], reverse=True)
    final_content_sigmoid_output1 = content_sigmoid_output1[1:11]
    index_final_output = [i[0] for i in final_content_sigmoid_output1]
    return df_shared_articles['title'].iloc[index_final_output]
    
 

# note the input given in this function is title+text

In [46]:
content_based_10_recommendations = gives_results(input_,sigmoid_kernel_output)

NameError: name 'input_' is not defined

In [ ]:
print(content_based_10_recommendations)

# Content Collaborative Filtering

In [46]:
#Creating a sparse pivot table with users in rows and items in columns
df_users_items_pivot_matrix = df_train_interactions.pivot(index='personId', 
                                                          columns='contentId', 
                                                          values='eventStrength').fillna(0)

df_users_items_pivot_matrix.head(10)

contentId             -9222795471790223670  -9216926795620865886  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   3.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9194572880052200111  -9192549002213406534  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9190737901804729417  -9189659052158407108  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9176143510534135851  -9172673334835262304  \
personId                                                           
-9223121837663643404                   0.0                   0.0   
-9212075797126931087                   0.0                   0.0   
-9207251133131336884                   0.0                   0.0   
-9199575329909162940                   0.0                   0.0   
-9196668942822132778                   0.0                   0.0   
-9188188261933657343                   0.0                   0.0   
-9172914609055320039                   0.0                   0.0   
-9156344805277471150                   0.0                   0.0   
-9120685872592674274                   0.0                   0.0   
-9109785559521267180                   0.0                   0.0   

contentId             -9171475473795142532  -9166778629773133902  ...  \
personId                                                          ...   
-9223121837663643404                   0.0                   0.0  ...   
-9212075797126931087                   0.0                   0.0  ...   
-9207251133131336884                   0.0                   0.0  ...   
-9199575329909162940                   0.0                   0.0  ...   
-9196668942822132778                   0.0                   0.0  ...   
-9188188261933657343                   0.0                   0.0  ...   
-9172914609055320039                   0.0                   0.0  ...   
-9156344805277471150                   0.0                   0.0  ...   
-9

In [47]:
from scipy.sparse import csr_matrix

df_to_csr_matrix = csr_matrix(df_users_items_pivot_matrix.values)

from sklearn.neighbors import NearestNeighbors


model_knn = NearestNeighbors(metric = 'cosine', algorithm = 'brute')
model_knn.fit(df_to_csr_matrix)

NearestNeighbors(algorithm='brute', leaf_size=30, metric='cosine',
                 metric_params=None, n_jobs=None, n_neighbors=5, p=2,
                 radius=1.0)

In [52]:
query_index = np.random.choice(df_users_items_pivot_matrix.shape[0])
print(query_index)
distances, indices = model_knn.kneighbors(df_users_items_pivot_matrix.iloc[query_index,:].values.reshape(1, -1), n_neighbors = 10)

855


In [62]:
for i in range(0, len(distances.flatten())):
    if i == 0:
        print('Recommendations for {0}:\n'.format(df_users_items_pivot_matrix.index[query_index]))
    else:
        print('{0}: {1}, with distance of {2}:'.format(i, df_users_items_pivot_matrix.index[indices.flatten()[len(distances.flatten()) - i]], distances.flatten()[len(distances.flatten()) - i]))

Recommendations for 4795304184119033503:

1: 8813266398846460512, with distance of 0.8226402204801077:
2: 1097401149118884075, with distance of 0.8210835074823184:
3: 6767912612638555664, with distance of 0.8195192810788916:
4: 899165159624701221, with distance of 0.8062877650548972:
5: -3390049372067052505, with distance of 0.8058393618540216:
6: -3325508780389598615, with distance of 0.8022932781043189:
7: 1929981821863611546, with distance of 0.8010202017055924:
8: -48161796606086482, with distance of 0.7983350149053692:
9: -9207251133131336884, with distance of 0.7818397866707592:
10: 1895326251577378793, with distance of 0.7813001183992603:
11: -1032019229384696495, with distance of 0.7797297249373938:
12: 3276989273780561693, with distance of 0.7499185800336661:
13: -5868110530814399805, with distance of 0.7451619458554203:
14: -8854674432071487111, with distance of 0.7415101454065283:
15: 1120069409160402054, with distance of 0.7144566297624038:
16: 3631158903888752348, with dis

# Pearson Correlation


In [112]:
query = ?

In [113]:

bones_ratings = df_users_items_pivot_matrix[query]
similar_to_bones = df_users_items_pivot_matrix.corrwith(bones_ratings)
corr_bones = pd.DataFrame(similar_to_bones, columns=['pearsonR'])
corr_bones.dropna(inplace=True)
corr_bones.sort_values('pearsonR', ascending=False).head(10)

pearsonR
contentId                     
-9222795471790223670  1.000000
-7174595923544719736  0.916198
 6519443272707698315  0.766385
 5152069678055228801  0.739172
 8871666435351703640  0.702023
-1737937277055036780  0.702023
-7454907763335725103  0.702023
 3465127447598546258  0.702023
 3218993154926038233  0.702023
 1362723651715193305  0.702023